README.md
# tradebots
Coming Soon


CHANGES
0.0.1 01/19/2019 -- Beginning of Development.


test/test_tradebots.py
"""
==========================================
TradeBots
Version: 0.0.1
Connectors Test Module
==========================================

Authors: Connor Sanders
"""

from tradebots import BollingerBot


# Test Connection here
api_token = '5ea8259e0844803f4a33c8661eb73c5c-9dd0b4290bd9ae29536246d672374fdb'
acct_id = '101-001-10278207-001'
symb = 'EUR_USD'
new_bollinger_bot = BollingerBot('oanda', api_token, acct_id, symb)


def check_candle_sticks():
    print('candle stick dict')
    candle_sticks_dict = new_bollinger_bot.candle_sticks(price='B', start='2019-01-03T21:48:55.000000000Z',
                                                         granularity='M10', end='2019-01-23T21:59:55.000000000Z')
    candle_sticks = candle_sticks_dict['candles']
    i = 0
    for candle_stick in candle_sticks:
        i += 1
        print(i)
        print(candle_stick.get('time', None))


def check_bollinger_bands():
    print('bollinger band check')
    granularity = 'M10'
    trained_bot = new_bollinger_bot.train_bot(granularity)
    #print(bolinger_bands)


check_bollinger_bands()

tradebots/__init__.py
"""
==========================================
TradeBots
Version: 0.0.1
Main Module
==========================================

Authors: Connor Sanders
"""

import numpy as np
from tradebots.connectors.oanda_connector import OandaConnector
from tradebots.connectors.alpaca_connector import AlpacaConnector
from tradebots.neutralnetworks.neural_network import NeuralNetwork


# Class that creates a new Bollinger Bot
class BollingerBot(object):

    def __init__(self, trade_platform, secret_token, account_id, symbol, practice=True):
        self.trade_platform = trade_platform
        self.secret_token = secret_token
        self.account_id = account_id
        self.symbol = symbol
        self.practice = practice
        if trade_platform == 'oanda':
            self.connection = OandaConnector(self.secret_token, self.account_id, self.symbol, self.practice)
        elif trade_platform == 'alpaca':
            self.connection = AlpacaConnector(self.secret_token, self.symbol)

    def current_price(self):
        return self.connection.current_pricing_data()

    def candle_sticks(self, price=None, granularity=None, start=None, end=None, smoothed=None, count=None):
        return self.connection.historical_candle_sticks(price, granularity, start, end, smoothed, count)

    def bollinger_bands(self, granularity, current=True, start=None, end=None):
        return self.connection.bollinger_band_data(granularity, current, start, end)

    def train_bot(self, granularity):
        bollinger_band_data = self.bollinger_bands(granularity)
        #print(bollinger_band_data)
        data_array = bollinger_band_data.values.T.tolist()
        data_array2 = bollinger_band_data.values.tolist()

        #print(data_array)
        #data_matrix = [list(l) for l in zip(*bollinger_band_data.values)]
        check_data_inputs = []
        i = 0
        for data in data_array2:
            cur_list = [data[1], data[4], data[5], data[6], data[7], data[8], data[9], data[10]]
            #print(cur_list)
            check_data_inputs.append(cur_list)
        # all variables except dates and prices
        # All close prices
        result_data = []
        for res_data in data_array[1]:
            result_data.append([res_data])
        data_results = np.array(result_data)
        data_inputs = np.array(check_data_inputs)
        print(data_results)
        print(data_inputs)
        nn = NeuralNetwork(data_inputs, data_results)
        for c in range(1500):
            nn.feedforward()
            nn.backprop()
        print(nn.output)
        return nn


tradebots/utils/math_utils.py
"""
==========================================
TradeBots
Version: 0.0.1
Math Utility Functions
==========================================

Authors: Connor Sanders
"""

import matplotlib.pyplot as plt
import math
import pandas as pd
import numpy as np


# Base Interval Mapping Dict
base_interval_scales_mapping = {
    "seconds": 0,
    "minutes": 60,
    "hours": 3600,
    "days": 86400,
    "weeks": 604800,
    "months": 2629800,
    "years": 31556926
}


# Function to calculate the scaled interval factor
def calc_scaled_interval_factor(interval, scale):
    base_scalar = base_interval_scales_mapping[scale]
    if base_scalar == 0:
        return int(interval)
    else:
        return base_scalar * int(interval)


# Function to calculate number of periods in 20 day span
def calc_num_periods(interval, scale):
    base_scalar = base_interval_scales_mapping[scale]
    day_seconds = base_interval_scales_mapping['days']
    if base_scalar == 0:
        return (day_seconds / int(interval)) * 20
    else:
        return (day_seconds / (base_scalar * int(interval))) * 20


# Function to calculate number of expected observations
def calc_num_of_observations(start, end, interval, scale):
    duration = end - start
    scaled_interval = calc_scaled_interval_factor(interval, scale)
    duration_in_s = duration.total_seconds()
    return [int(divmod(duration_in_s, scaled_interval)[0]), scaled_interval]


# Function to calculate stock's RSI
def calc_rsi(price):
    avg_gain = price[price > 0].sum() / n
    avg_loss = -price[price < 0].sum() / n
    rs = avg_gain / avg_loss
    return 100 - 100 / (1 + rs)


def calc_rsi_values(candlestick_dicts, n=14):
    df = pd.DataFrame(candlestick_dicts)
    df['rsi'] = (df['close'] - df['close'].shift(1)).fillna(0)
    return pd.rolling_apply(df, n, calc_rsi)


# Function to calculate bollinger values
def calc_bollinger_values(candlestick_dicts, interval, time_scale):
    #print(candlestick_dicts)
    df = pd.DataFrame(candlestick_dicts)
    if time_scale == 'days':
        df['moving_average'] = df['close'].rolling(window=45).mean()
    else:
        num_periods = int(calc_num_periods(interval, time_scale))
        print(num_periods)
        delta = df.close.diff()
        up_days = delta.copy()
        up_days[delta <= 0] = 0.0
        down_days = abs(delta.copy())
        down_days[delta > 0] = 0.0
        RS_up = up_days.rolling(window=17).mean()
        RS_down = down_days.rolling(window=17).mean()
        df['rsi'] = 100 - 100 / (1 + RS_up / RS_down)
        df['moving_average'] = df['close'].rolling(window=200).mean()
        df['standard_deviation'] = df['close'].rolling(window=200).std()
        df['upper_band'] = df['moving_average'] + (df['standard_deviation'] * 2)
        df['lower_band'] = df['moving_average'] - (df['standard_deviation'] * 2)
        df['band_diff'] = df['upper_band'] - df['lower_band']
    df = df.set_index('time')
    #df[['close', 'moving_average', 'upper_band', 'lower_band']].plot(figsize=(24, 12))
    #plt.title('30 Day Bollinger Band Chart')
    #plt.ylabel('Price (USD)')
    #plt.show()
    return df.dropna()


# Function that calculates a
def sigmoid(x):
    return 1.0 / (1 + np.exp(-x))


# Function that calculates the derivative of a sigmoid
def sigmoid_derivative(x):
    return x * (1.0 - x)


tradebots/utils/request_utils.py
"""
==========================================
TradeBots
Version: 0.0.1
Request Utility Functions
==========================================

Authors: Connor Sanders
"""

import sys
import json
import time
if sys.version_info < (3, 0, 0):
    from urllib2 import Request, urlopen
else:
    from urllib.request import Request, urlopen


# Function to handle a get request
def base_request(url, req_type, headers, req_data=None):
    if req_type != 'GET' and req_data is not None:
        req = Request(url, data=json.dumps(req_data).encode('utf8'))
    else:
        req = Request(url)
    if headers is not None:
        for header_key, header_val in headers.items():
            req.add_header(header_key, header_val)
    if req_type != 'GET':
        req.get_method = lambda: req_type
    return urlopen(req)


# Function to handle a new request
def handle_request(url, req_type, headers, req_data=None, i=0):
    if i == 4:
       return None
    try:
        if req_data is None:
            res = base_request(url, req_type, headers)
        else:
            res = base_request(url, req_type, headers, req_data)
        if res.code == 500:
            i += 1
            return handle_request(url, req_type, headers, req_data, i)
        else:
            return res
    except:
        i += 1
        return handle_request(url, req_type, headers, req_data, i)


# Function to handle a get request
def get_request(url, headers=None):
    return handle_request(url, 'GET', headers)


# Function to handle a post request
def post_request(url, req_data, headers=None):
    return handle_request(url, 'POST', headers, req_data)


# Function to handle a put request
def put_request(url, req_data, headers=None):
    return handle_request(url, 'PUT', headers, req_data)


# Function to handle a put request
def patch_request(url, req_data, headers=None):
    return handle_request(url, 'PATCH', headers, req_data)


# Function to handle a delete request
def delete_request(url, headers=None):
    return handle_request(url, 'DELETE', headers)


tradebots/utils/datetime_utils.py
"""
==========================================
TradeBots
Version: 0.0.1
DateTime Utility Functions
==========================================

Authors: Connor Sanders
"""

import datetime


# Function to calculate datetime map
def generate_dt_range_map(num_of_requests, seconds_per_request, end_dt, start_dt):
    date_time_ranges = []
    i = 0
    cur_start = start_dt
    while i < num_of_requests:
        cur_end = cur_start + datetime.timedelta(0, seconds_per_request)
        i += 1
        if i < num_of_requests:
            cur_time_range_list = [cur_start, cur_end]
        else:
            cur_time_range_list = [cur_start, end_dt]
        date_time_ranges.append(cur_time_range_list)
        cur_start = cur_end + datetime.timedelta(0, int(seconds_per_request / 3000))
    return date_time_ranges


tradebots/connectors/alpaca_connector.py
"""
==========================================
TradeBots
Version: 0.0.1
Alpaca Connector Module
==========================================

Authors: Connor Sanders and Daniel Pivalizza
"""

import sys


# Class that instantiates a new Alpaca Connection
class AlpacaConnector(object):

    # Initialize an Alpaca Connection
    def __init__(self, secret_token, symbol):

        # Function that checks if token is valid
        def check_token(secret_token):
            if secret_token:
                # Write code here
                return secret_token
            print('Token is not recognized by Alpaca')
            sys.exit(1)

        # Function that check if symbol is valid upon bot instantiation
        def check_symbol(symbol):
            if symbol:
                # Write code here
                return symbol
            print('Alpaca Connector does not recognize the symbol, ' + str(symbol))
            sys.exit(1)

        self.secret_token = check_token(secret_token)
        self.symbol = check_symbol(symbol)

    # Public Method that returns a dictionary of current candle stick values
    def current_candle_sticks(self):
        pass

    # Public Method that returns a dictionary of current bollinger band values
    def bollinger_bands(self):
        pass

    # Public Method that executes an open trade order
    def open_trade(self):
        pass

    # Public Method that executes a close trade order
    def close_trade(self):
        pass


tradebots/neutralnetworks/neural_network.py
"""
==========================================
TradeBots
Version: 0.0.1
Neural Network Module
==========================================

Authors: Connor Sanders
"""

import numpy as np
import tradebots.utils.math_utils as math_utils


class NeuralNetwork:

    def __init__(self, x, y):
        self.input = x
        self.weights1 = np.random.rand(self.input.shape[1], 2608)
        self.weights2 = np.random.rand(2608, 1)
        self.y = y
        self.output = np.zeros(y.shape)

    def feedforward(self):
        self.layer1 = math_utils.sigmoid(np.dot(self.input, self.weights1))
        self.output = math_utils.sigmoid(np.dot(self.layer1, self.weights2))

    def backprop(self):
        # application of the chain rule to find derivative of the loss function with respect to weights2 and weights1
        d_weights2 = np.dot(self.layer1.T, (2*(self.y - self.output) * math_utils.sigmoid_derivative(self.output)))
        d_weights1 = np.dot(self.input.T,  (np.dot(2*(self.y - self.output) * math_utils.sigmoid_derivative(self.output), self.weights2.T) * math_utils.sigmoid_derivative(self.layer1)))

        # update the weights with the derivative (slope) of the loss function
        self.weights1 += d_weights1
        self.weights2 += d_weights2


'''
if __name__ == "__main__":
    x = np.array([[0.2,0,1.2],
                  [0.2,1.1,1],
                  [1,0,1],
                  [1,1,1]])
    y = np.array([[0],[1],[1],[0]])
    nn = NeuralNetwork(x, y)

    print(nn.weights1)
    print(nn.weights2)
    for i in range(1500):
        nn.feedforward()
        nn.backprop()
    print(nn.output)
    print(nn.weights1)
    print(nn.weights2)
'''

tradebots/connectors/oanda_connector.py
"""
==========================================
TradeBots
Version: 0.0.1
Oanda Connector Module
==========================================

Authors: Connor Sanders
"""

import sys
import json
import datetime
import tradebots.utils.request_utils as request_utils
import tradebots.utils.math_utils as math_utils
import tradebots.utils.datetime_utils as datetime_utils


# Oanda Scale Mapping
time_scaling_map = {
    'S': 'seconds',
    'M': 'minutes',
    'H': 'hours',
    'D': 'days',
    'W': 'weeks',
    'M1': 'months'
}


# Class that instantiates a new Oanda Connection
class OandaConnector(object):

    # Initialize an Oanda Connection Object
    def __init__(self, secret_token, account_id, symbol, practice):

        # Function that checks if token is valid
        def check_settings(check_account_id, base_url, base_headers, check_type, check_symbol=None):
            if check_type == 'accounts':
                get_url = base_url + '/v3/accounts'
                check_key = 'id'
                check_param = check_account_id
            else:
                get_url = base_url + '/v3/accounts/' + check_account_id + '/instruments'
                check_key = 'name'
                check_param = check_symbol
            try:
                get_req = request_utils.get_request(get_url, base_headers)
                if get_req.code != 200:
                    print('setup error: Token is not recognized by Oanda')
                    sys.exit(1)
                else:
                    res_list = json.loads(get_req.read().decode('utf8')).get(check_type, None)
                    if res_list is not None and len(res_list) > 0:
                        for res_obj in res_list:
                            if res_obj.get(check_key, None) == check_param:
                                return res_obj
                    print(str(check_type) + ' setup error: ' + str(check_param) + ' is not recognized by Oanda')
                    sys.exit(1)
            except:
                print('setup error: Token is not recognized by Oanda')
                sys.exit(1)

        # Object Initialization Variables
        self.practice = practice
        self.base_headers = {"Authorization": 'Bearer ' + str(secret_token)}
        if practice is True:
            self.base_url = 'https://api-fxpractice.oanda.com'
            self.base_streaming_url = 'https://stream-fxpractice.oanda.com/'
        else:
            self.base_url = 'https://api-fxtrade.oanda.com'
            self.base_streaming_url = 'https://stream-fxtrade.oanda.com/'
        self.account_id = account_id
        self.secret_token = secret_token
        check_settings(self.account_id, self.base_url, self.base_headers, 'accounts')
        self.symbol = symbol.replace('/', '_')
        self.symbol_dict = check_settings(self.account_id, self.base_url, self.base_headers, 'instruments', self.symbol)
        self.granularity_meta = {''}

    # Private Static Method to generate historical candle stick req URL
    @staticmethod
    def _generate_candle_stick_url(base_get_url, price, granularity, start, end, smoothed, count):
        i = 0
        if price is not None:
            base_get_url += '?price=' + str(price)
            i += 1
        if granularity is not None:
            if i == 0:
                base_get_url += '?granularity=' + str(granularity)
            else:
                base_get_url += '&granularity=' + str(granularity)
            i += 1
        if count is not None:
            if i == 0:
                base_get_url += '?count=' + str(count)
            else:
                base_get_url += '&count=' + str(count)
            i += 1
        if start is not None:
            if i == 0:
                base_get_url += '?from=' + str(start).replace(':', '%3A')
            else:
                base_get_url += '&from=' + str(start).replace(':', '%3A')
            i += 1
        if end is not None:
            base_get_url += '&to=' + str(end).replace(':', '%3A')
        if smoothed is not None:
            if i == 0:
                base_get_url += '?smoothed=' + str(smoothed)
            else:
                base_get_url += '&smoothed=' + str(smoothed)
        return base_get_url

    # Private static method to return Oanda's time scale factors
    @staticmethod
    def _get_time_scales(granularity):
        scale = "".join([i for i in granularity if i.isalpha()])
        interval = "".join([i for i in granularity if i.isdigit()])
        time_scale = time_scaling_map[scale]
        return [interval, time_scale]

    # Public Method that returns a dictionary of historical candle stick values
    def historical_candle_sticks(self, price=None, granularity=None, start=None, end=None, smoothed=None, count=None):
        start_dt = datetime.datetime.strptime(start.split('.')[0] + 'Z', '%Y-%m-%dT%H:%M:%SZ')
        end_dt = datetime.datetime.strptime(end.split('.')[0] + 'Z', '%Y-%m-%dT%H:%M:%SZ')
        time_scale_dict = self._get_time_scales(granularity)
        interval = time_scale_dict[0]
        time_scale = time_scale_dict[1]
        num_observations_data = math_utils.calc_num_of_observations(start_dt, end_dt, interval, time_scale)
        num_observations = num_observations_data[0]
        scaled_interval = num_observations_data[1]
        seconds_per_request = int(scaled_interval * 3000)
        num_of_requests = int(num_observations / 3000)
        date_ranges_list = datetime_utils.generate_dt_range_map(num_of_requests, seconds_per_request, end_dt, start_dt)
        base_get_url = self.base_url + '/v3/instruments/' + self.symbol + '/candles'
        appended_re_candle_data = []
        re_data = {}
        for datetime_list in date_ranges_list:
            c_start = datetime_list[0].isoformat("T") + "Z"
            c_end = datetime_list[1].isoformat("T") + "Z"
            get_url = self._generate_candle_stick_url(base_get_url, price, granularity, c_start, c_end, smoothed, count)
            try:
                get_req = request_utils.get_request(get_url, self.base_headers)
                if get_req:
                    if get_req.code != 200:
                        appended_re_candle_data.append({c_start + ' to ' + c_end: False})
                    else:
                        re_data = json.loads((get_req.read().decode('utf8')))
                        if len(appended_re_candle_data) > 0:
                            appended_re_candle_data += re_data['candles']
                        else:
                            appended_re_candle_data = re_data['candles']
            except:
                appended_re_candle_data.append({c_start + ' to ' + c_end: None})
        re_data['candles'] = appended_re_candle_data
        return re_data

    # Public Method that returns a dictionary of current pricing data
    def current_pricing_data(self):
        get_url = self.base_url + '/v3/accounts/' + self.account_id + '/pricing?instruments=' + self.symbol
        try:
            get_req = request_utils.get_request(get_url, self.base_headers)
            if get_req.code != 200:
                return False
            return json.loads(get_req.read())
        except:
            return False

    # Private Method that formats returned candle stick pricing data
    @staticmethod
    def _format_candle_stick_data(raw_candle_sticks, price='mid'):
        raw_candle_stick_prices = raw_candle_sticks['candles']
        candle_stick_prices = []
        for raw_candle_stick in raw_candle_stick_prices:
            formated_candle_stick_obj = {
                'volume': float(raw_candle_stick['volume']),
                'time': raw_candle_stick['time'],
                'high': float(raw_candle_stick[price].get('h', None)),
                'close': float(raw_candle_stick[price].get('c', None)),
                'open': float(raw_candle_stick[price].get('o', None)),
                'low': float(raw_candle_stick[price].get('l', None)),
            }
            candle_stick_prices.append(formated_candle_stick_obj)
        raw_candle_sticks.update({'candles': candle_stick_prices})
        return raw_candle_sticks

    # Public Method to get RSI data
    @staticmethod
    def rsi_data(candlesticks_dataframe):
        return math_utils.calc_rsi_values(candlesticks_dataframe)

    # Public Method that returns a dictionary of current bollinger band values
    def bollinger_band_data(self, granularity, current=True, start=None, end=None):
        if current is True:
            # get last 20 days worth of candle stick data for submitted granularity; if less than M10, run multiple pulls
            end_dt = datetime.datetime.utcnow()
            start_dt = end_dt - datetime.timedelta(days=22)
            end = end_dt.isoformat("T") + "Z"
            start = start_dt.isoformat("T") + "Z"
            raw_candle_sticks = self.historical_candle_sticks(None, granularity, start, end)
            candle_sticks = self._format_candle_stick_data(raw_candle_sticks)
            #TODO Calculate Bollinger Bands
            time_scale_dict = self._get_time_scales(granularity)
            interval = time_scale_dict[0]
            time_scale = time_scale_dict[1]
            return math_utils.calc_bollinger_values(candle_sticks['candles'], interval, time_scale)
        else:
            pass
        pass

    # Public Method that executes an open trade order
    def open_trade(self):
        pass

    # Public Method that executes a close trade order
    def close_trade(self):
        pass
